beacon-chain/blockchain/service.go

```diff
@@ -236,120 +220,81 @@ func (c *ChainService) blockProcessing(done <-chan struct{}) {
-			var canProcessBlock bool
-			var canProcessAttestations bool
-			var err error
-			var blockVoteCache map[[32]byte]*types.VoteCache
-
-			h, err := block.Hash()
-			if err != nil {
-				log.Debugf("Could not hash incoming block: %v", err)
-			}
-
-			receivedSlotNumber := block.SlotNumber()
-
-			log.WithField("blockHash", fmt.Sprintf("0x%x", h)).Info("Received full block, processing validity conditions")
-
-			parentExists, err := c.chain.hasBlock(block.ParentHash())
+			// 1. Validate the block
+			// 2. If a candidate block with a lower slot exists, run the fork choice rule
+			// 3. Save the block
+			// 4. If a candidate block exists, exit
+			// 4. Calculate the active and crystallized state for the block
+			// 5. Set the block as the new candidate block
+			aState := c.chain.ActiveState()
+			cState := c.chain.CrystallizedState()
+			blockHash, err := block.Hash()
 			if err != nil {
-				log.Debugf("Could not check existence of parent hash: %v", err)
-			}
-
-			// If parentHash does not exist, received block fails validity conditions.
-			if !parentExists && receivedSlotNumber > 1 {
+				log.Errorf("Failed to get hash of block: %v", err)
 				continue
 			}
 
 			// Process block as a validator if beacon node has registered, else process block as an observer.
-			canProcessBlock, err = c.chain.CanProcessBlock(c.web3Service.Client(), block, c.isValidator)
+			parentExists, err := c.chain.hasBlock(block.ParentHash())
 			if err != nil {
-				// We might receive a lot of blocks that fail validity conditions,
-				// so we create a debug level log instead of an error log.
-				log.Debugf("Incoming block failed validity conditions: %v", err)
-			}
-
-			// Process attestations as a beacon chain node.
-			var processedAttestations []*pb.AttestationRecord
-			for index, attestation := range block.Attestations() {
-				if err := c.chain.processAttestation(index, block); err != nil {
-					canProcessAttestations = false
-					log.Errorf("could not process attestation for block %d because %v", block.SlotNumber(), err)
-				} else {
-					canProcessAttestations = true
-					processedAttestations = append(processedAttestations, attestation)
-				}
-			}
-
-			// If we cannot process an attestation in this block, we keep listening.
-			if !canProcessAttestations {
+				log.Errorf("Could not check existence of parent: %v", err)
 				continue
 			}
-			// If we cannot process this block, we keep listening.
-			if !canProcessBlock {
+			if !parentExists {
+				log.Errorf("parent does not exist: %x", block.ParentHash())
 				continue
 			}
 
-			// With a valid beacon block, we can compute its attestations and store its votes/deposits in cache.
-			for index := range block.Attestations() {
-				blockVoteCache, err = c.chain.calculateBlockVoteCache(index, block)
-				if err != nil {
-					log.Debugf("could not calculate new block vote cache: %v", nil)
-				}
+			if !parentExists || !c.doesPoWBlockExist(block) || !block.IsValid(aState, cState) {
```

Comment:
checking again for `!parentExists` is redundant here since you checked for it above

Reply:
ah, nice catch. thanks!